File: go/binary_search/find_first.go
package binary_search

// firstInt should return first index of val in a sorted array arr, or -1 if not found
func FindFirstInt(arr []int, val int) int {
	l, r := 0, len(arr)
	for l < r {
		if m := (l + r) / 2; arr[m] < val {
			l = m + 1
		} else {
			r = m
		}
	}

	if l < len(arr) && arr[l] == val {
		return l
	}
	return -1
}

File: python/binary_search/find_first.py

# find_first_int should return first index of val in a sorted array arr, or -1 if not found
def find_first_int(arr: list, val: int) -> bool:
	# raise NotImplementedError('TODO')

	l, r = 0, len(arr)
	while l < r:
		m = (l + r) // 2
		if arr[m] < val:
			l = m+1
		else:
			r = m

	return l if l < len(arr) and arr[l] == val else -1
